src/nendo_plugin_quantize_core/plugin.py

```diff
@@ -34,13 +40,97 @@ class CoreQuantizer(NendoGeneratePlugin):
     config: NendoConfig = None
     logger: Logger = None
 
+    def sequence_generator(self, duration: int):
+        """Generates a sequence of powers of 2 until the duration.
+
+        Args:
+            duration: The maximum duration.
+
+        Yields:
+            An int which is the next power of 2 in the sequence.
+        """
+        power = int(np.log2(duration))
+        yield from 2 ** np.arange(power, 0, -1)
+
+    def extract_beat(self, y, sr):
+        """Extracts beat from a given audio signal and converts beats to frames.
+
+        Args:
+            y: The audio signal.
+            sr: The sample rate.
+
+        Returns:
+            A tuple of (tempo, beat_frames).
+        """
+        _, y_percussive = librosa.effects.hpss(y)
+
+        # Beat Track: Simple
+        tempo, beats = librosa.beat.beat_track(y=y_percussive, sr=sr, trim=False)
+
+        # Beat track: Complex
+        # tempo, beats = librosa.beat.beat_track(
+        #     sr=sr, onset_envelope=librosa.onset.onset_strength(y=y_percussive, sr=sr), trim=False)
+
+        beat_frames = librosa.frames_to_samples(beats)
+
+        return tempo, beat_frames
+
+    def extract_beat_essentia(self, y, sr):
+        """Extracts beat from a given audio signal and converts beats to frames.
+
+        Args:
+            y: The audio signal.
+            sr: The sample rate.
+
+        Returns:
+            A tuple of (tempo, beat_frames).
+        """
+        rhythm_extractor = es.RhythmExtractor2013(method="multifeature")
+        bpm, beats, beats_confidence, _, _ = rhythm_extractor(y)
+
+        # Convert beats to frames
+        # (Essentia's RhythmExtractor2013 returns beats in seconds)
+        beat_frames = (beats * sr).astype(int)
+
+        # Ensure unique beat frames
+        beat_frames = np.unique(beat_frames)
+
+        # Ensure monotonic beat frames
+        beat_frames_diff = np.diff(beat_frames)
+        if np.any(beat_frames_diff <= 0):
+            beat_frames = beat_frames_diff.cumsum()
+
+        return bpm, beat_frames
+
+    def construct_time_map(
+        self,
+        beat_frames: np.ndarray,
+        scale: float,
+        audio_length: int,
+    ) -> List[List[int]]:
+        """Constructs the time map using Numpy.
+
+        Args:
+            beat_frames: Array of beat frames.
+            scale: Scaling factor.
+            audio_length: Length of the audio in samples.
+
+        Returns:
+            The constructed time map.
+        """
+        beat_frames_scaled = np.round(beat_frames * scale).astype(int)
+        time_map = np.column_stack((beat_frames, beat_frames_scaled)).tolist()
+        time_map.append([audio_length, int(audio_length * scale)])
+
+        return time_map
+
     @NendoGeneratePlugin.run_track
-    def core_quantize(
+    def quantize_audio(
         self,
         track: NendoTrack,
         bpm: int = 120,
         keep_original_bpm: bool = settings.keep_original_bpm,
-    ):
+    ) -> Tuple[float, float]:
```

Comment:
```suggestion
    ) -> NendoTrack:
```

does return a `NendoTrack`